common/archiver/s3store/historyArchiver.go

```diff
@@ -168,26 +178,61 @@ func (h *historyArchiver) Archive(
 			return archiver.ErrHistoryMutated
 		}
 
-		historyBatches = append(historyBatches, historyBlob.Body...)
-	}
+		encodedHistoryPart, err := encode(historyBlob.Body)
+		if err != nil {
+			logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
+			return err
+		}
 
-	encodedHistoryBatches, err := encode(historyBatches)
-	if err != nil {
-		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
-		return err
+		key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion, progress.BatchIdx)
+		if err := upload(ctx, h.s3cli, URI, key, encodedHistoryPart); err != nil {
+			logger := logger.WithTags(tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
+			if isRetryableError(err) {
+				logger.Error(archiver.ArchiveTransientErrorMsg)
+			} else {
+				logger.Error(archiver.ArchiveNonRetriableErrorMsg)
+			}
+			return err
+		}
+		uploadedSize := int64(binary.Size(encodedHistoryPart))
+		scope.AddCounter(metrics.HistoryArchiverTotalUploadSize, uploadedSize)
+		scope.AddCounter(metrics.HistoryArchiverHistorySize, uploadedSize)
+		progress.BatchIdx = progress.BatchIdx + 1
+		err = saveHistoryIteratorState(ctx, featureCatalog, historyIterator, &progress)
+		if err != nil {
+			logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
+			return err
+		}
 	}
 
-	key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion)
+	scope.IncCounter(metrics.HistoryArchiverArchiveSuccessCount)
+	return nil
+}
 
-	if err := upload(ctx, h.s3cli, URI, key, encodedHistoryBatches); err != nil {
-		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
-		return err
+func loadHistoryIterator(ctx context.Context, request *archiver.ArchiveHistoryRequest, historyManager persistence.HistoryManager, featureCatalog *archiver.ArchiveFeatureCatalog, progress *uploadProgress) (historyIterator archiver.HistoryIterator, err error) {
+	if featureCatalog.ProgressManager != nil {
+		if featureCatalog.ProgressManager.HasProgress(ctx) {
+			err = featureCatalog.ProgressManager.LoadProgress(ctx, progress)
+			if err != nil {
+				return nil, err
```

Comment:
Since the function will never return an error, I think we can remove err from the function signature and the logic at the caller side for checking the error.

If we failed to `LoadProgress`, you can just call `NewHistoryIterator` which won't an error.

Reply:
Great, fixed!